cardano-api/src/Cardano/Api/LocalStateQuery.hs

```diff
@@ -0,0 +1,190 @@
+                  (nodeToClientProtocolVersion, supportedNodeToClientVersions)
+import           Ouroboros.Consensus.Node.ProtocolInfo (ProtocolInfo(..))
+import           Ouroboros.Consensus.Node.Run (RunNode, nodeNetworkMagic)
+import           Ouroboros.Consensus.Shelley.Ledger
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto (TPraosCrypto)
+
+import           Ouroboros.Network.Block (Point)
+import           Ouroboros.Network.Mux
+                   ( AppType(..), OuroborosApplication(..),
+                     MuxPeer(..), RunMiniProtocol(..))
+import           Ouroboros.Network.NodeToClient (ConnectionId, DictVersion, Handshake,
+                   LocalAddress, LocalConnectionId, NetworkConnectTracers (..),
+                   NodeToClientProtocols (..), NodeToClientVersion, NodeToClientVersionData (..),
+                   TraceSendRecv, Versions)
+import qualified Ouroboros.Network.NodeToClient as NodeToClient
+import           Ouroboros.Network.Protocol.LocalStateQuery.Client
+                   (ClientStAcquired (..), ClientStAcquiring (..), ClientStIdle (..),
+                    ClientStQuerying (..), LocalStateQueryClient(..), localStateQueryClientPeer)
+import           Ouroboros.Network.Protocol.LocalStateQuery.Type (AcquireFailure (..))
+
+import qualified Shelley.Spec.Ledger.PParams as Ledger (PParams)
+
+-- | Query the current protocol parameters from a Shelley node via the local
+-- state query protocol.
+queryPParamsFromLocalState
+  :: TPraosCrypto c
+  => Protocol (ShelleyBlock c) (BlockProtocol (ShelleyBlock c))
+  -> SocketPath
+  -> Point (ShelleyBlock c)
+  -> ExceptT AcquireFailure IO Ledger.PParams
+queryPParamsFromLocalState ptcl socketPath point = do
+  let pointAndQuery = (point, GetCurrentPParams)
+  pParamsVar <- liftIO newEmptyTMVarM
+  liftIO $ queryNodeLocalState
+    pParamsVar
+    nullTracer
+    ptcl
+    socketPath
+    pointAndQuery
+  newExceptT $ atomically $ takeTMVar pParamsVar
+
+-- | Establish a connection to a Shelley node and execute the provided query
+-- via the local state query protocol.
+queryNodeLocalState
+  :: forall blk result. RunNode blk
+  => StrictTMVar IO (Either AcquireFailure result)
+  -> Trace IO Text
+  -> Protocol blk (BlockProtocol blk)
+  -> SocketPath
+  -> (Point blk, Query blk result)
+  -> IO ()
+queryNodeLocalState resultVar trce ptcl (SocketFile socketPath) pointAndQuery = do
+    logInfo trce $ "queryNodeLocalState: Connecting to node via " <> textShow socketPath
+    let ProtocolInfo{pInfoConfig} = protocolInfo ptcl
+    NodeToClient.withIOManager $ \iocp -> do
+      NodeToClient.connectTo
+        (NodeToClient.localSnocket iocp socketPath)
+        NetworkConnectTracers
+          { nctMuxTracer = muxTracer
+          , nctHandshakeTracer = handshakeTracer
+          }
+        (localInitiatorNetworkApplication proxy resultVar trce pInfoConfig pointAndQuery)
+        socketPath
+  where
+    -- TODO: it should be passed as an argument
```

Comment:
I think this stuff is probably still in flux, so this is fine!

Reply:
I think I actually copied this comment along with some of your code from the `TxSubmit` module :smile: